src/domain/core/src/use_cases/view_dataset_use_case.rs

```diff
@@ -0,0 +1,80 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use internal_error::InternalError;
+use odf::dataset::DatasetNotFoundError;
+use thiserror::Error;
+
+use crate::auth::DatasetActionUnauthorizedError;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// TODO: Private Datasets: tests
+#[async_trait::async_trait]
+pub trait ViewDatasetUseCase: Send + Sync {
+    async fn execute(
+        &self,
+        dataset_ref: &odf::DatasetRef,
+    ) -> Result<odf::DatasetHandle, ViewDatasetUseCaseError>;
+
+    async fn execute_multi(
+        &self,
+        dataset_refs: Vec<odf::DatasetRef>,
+    ) -> Result<ViewMultiResult, InternalError>;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug)]
+pub struct ViewMultiResult {
+    pub viewable_resolved_refs: Vec<(odf::DatasetRef, odf::DatasetHandle)>,
+    pub inaccessible_refs: Vec<(odf::DatasetRef, NotAccessibleError)>,
+}
+
+impl ViewMultiResult {
+    pub fn into_inaccessible_input_datasets_message(self) -> String {
+        use itertools::Itertools;
+
+        let inaccessible_dataset_refs_it = self
+            .inaccessible_refs
+            .into_iter()
+            .map(|(dataset_ref, _)| dataset_ref);
+        let joined_inaccessible_datasets = inaccessible_dataset_refs_it
+            .map(|dataset_ref| format!("'{dataset_ref}'"))
+            .join(", ");
+
+        format!("Some input dataset(s) are inaccessible: {joined_inaccessible_datasets}")
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Error, Debug)]
+pub enum NotAccessibleError {
+    #[error(transparent)]
+    NotFound(#[from] DatasetNotFoundError),
+
+    #[error(transparent)]
+    Unauthorized(#[from] DatasetActionUnauthorizedError),
+}
+
+#[derive(Error, Debug)]
+pub enum ViewDatasetUseCaseError {
+    #[error(transparent)]
+    NotAccessible(#[from] NotAccessibleError),
```

Comment:
We have `odf::AccessError` used everywhere.
Also, why not to inline these varients into the main error? The benefit of this grouping?

Reply:
> We have `odf::AccessError` used everywhere.

Added to `ViewDatasetUseCaseError` as an enum variant

> Also, why not to inline these varients into the main error? The benefit of this grouping?

I was confused earlier that I can't put `DatasetNotFoundError` inside `AccessError` without doubts. 

As a result, one of the reasons why `DatasetNotFoundError` is added as `ViewDatasetUseCaseError::NotFound()`